[WIP] [RFC] Use smart_open to also support other than plain file storage

a17t opened the pull request:
Hello everyone,

I initially added this as an issue in the old paperless-ng repo. Again picked it up on our latest 1.6 release.

The main change is to use https://github.com/RaRe-Technologies/smart_open, which replaces the built-in "open" function and allows to use
`
    S3
    HTTP, HTTPS (read-only)
    SSH, SCP and SFTP
    WebHDFS
    GCS
    Azure Blob Storage
`
as a storage backend. 

Before merging this, I would like to add additional checks, as remote storage will not work under all conditions.

Currently I use it for my setup to use a S3 Bucked as Media and Data location (with postgresql database backend). 

Using S3 as data location will not work with sqlite (and I am sure there are much more situation this might not work).

Introducing this as an abstraction for opening files could also enable us to support things like #335 (Native Samba Storage support), by implementing corresponding backend for smart_open.

a17t committed:
Update classifier.py, consumer.py, and 12 more files...

a17t committed:
Update requirements.txt

stumpylog committed:
Updates the target branch for GHA updates to be dev

shamoon committed:
Merge pull request #804 from paperless-ngx/dependabot-gha-target-dev

Updates the target branch for GHA updates to be dev

qcasey committed:
Remove expected behavior section

qcasey committed:
Merge pull request #813 from paperless-ngx/remove-expected-behavior

Remove expected behavior section

qcasey committed:
Add back labels removed in 30834e

qcasey committed:
Add back review groups

qcasey committed:
Assign GHA bumps to CI/CD team

qcasey committed:
Fix comment directory

shamoon committed:
Merge pull request #829 from paperless-ngx/fix-missing-labels-dep

Fix `dependabot.yml` config

qcasey committed:
Chore: Improve Bug Report template (#859)

* Try to get architecture in BR

* Move screenshots to Description

* Add link to search

* Specify doc page

* Clarify bug report intro

* Add missing period

* Other missing periods for consistiency

* Add line breaks, architecture req

* Expand Currently...

* Removing trailing quote

dabenzel committed:
I have created a shortcut to Upload PDF files

//

dabenzel committed:
Update docs/scanners.rst

Co-authored-by: shamoon <4887959+shamoon@users.noreply.github.com>

qcasey committed:
Bump version

paperlessngx-bot committed:
New Crowdin updates (#816)

* New translations messages.xlf (Polish)
[ci skip]

* New translations messages.xlf (Italian)
[ci skip]

* New translations messages.xlf (Luxembourgish)
[ci skip]

* New translations messages.xlf (Portuguese, Brazilian)
[ci skip]

* New translations messages.xlf (Chinese Simplified)
[ci skip]

* New translations messages.xlf (Turkish)
[ci skip]

* New translations messages.xlf (Swedish)
[ci skip]

* New translations messages.xlf (Slovenian)
[ci skip]

* New translations messages.xlf (Slovak)
[ci skip]

* New translations messages.xlf (Portuguese)
[ci skip]

* New translations messages.xlf (Dutch)
[ci skip]

* New translations messages.xlf (Hebrew)
[ci skip]

* New translations messages.xlf (Danish)
[ci skip]

* New translations messages.xlf (Finnish)
[ci skip]

* New translations messages.xlf (Greek)
[ci skip]

* New translations messages.xlf (German)
[ci skip]

* New translations messages.xlf (Czech)
[ci skip]

* New translations messages.xlf (Belarusian)
[ci skip]

* New translations messages.xlf (Spanish)
[ci skip]

* New translations messages.xlf (Romanian)
[ci skip]

* New translations messages.xlf (Russian)
[ci skip]

* New translations messages.xlf (French)
[ci skip]

* New translations messages.xlf (Serbian (Latin))
[ci skip]

* New translations messages.xlf (Arabic, Saudi Arabia)
[ci skip]

* New translations messages.xlf (Belarusian)
[ci skip]

* New translations messages.xlf (Czech)
[ci skip]

* New translations messages.xlf (German)
[ci skip]

* New translations messages.xlf (Chinese Simplified)
[ci skip]

* New translations messages.xlf (Arabic, Saudi Arabia)
[ci skip]

* New translations messages.xlf (Polish)
[ci skip]

* New translations messages.xlf (Italian)
[ci skip]

* New translations messages.xlf (Luxembourgish)
[ci skip]

* New translations messages.xlf (Portuguese, Brazilian)
[ci skip]

* New translations messages.xlf (Chinese Simplified)
[ci skip]

* New translations messages.xlf (Turkish)
[ci skip]

* New translations messages.xlf (Swedish)
[ci skip]

* New translations messages.xlf (Slovenian)
[ci skip]

* New translations messages.xlf (Slovak)
[ci skip]

* New translations messages.xlf (Portuguese)
[ci skip]

* New translations messages.xlf (Dutch)
[ci skip]

* New translations messages.xlf (Hebrew)
[ci skip]

* New translations messages.xlf (Danish)
[ci skip]

* New translations messages.xlf (Finnish)
[ci skip]

* New translations messages.xlf (Greek)
[ci skip]

* New translations messages.xlf (German)
[ci skip]

* New translations messages.xlf (Czech)
[ci skip]

* New translations messages.xlf (Belarusian)
[ci skip]

* New translations messages.xlf (Spanish)
[ci skip]

* New translations messages.xlf (Romanian)
[ci skip]

* New translations messages.xlf (Russian)
[ci skip]

* New translations messages.xlf (French)
[ci skip]

* New translations messages.xlf (Serbian (Latin))
[ci skip]

* New translations messages.xlf (Arabic, Saudi Arabia)
[ci skip]

* New translations messages.xlf (German)
[ci skip]

* New translations messages.xlf (German)
[ci skip]

* New translations messages.xlf (German)
[ci skip]

* New translations messages.xlf (Serbian (Latin))
[ci skip]

* New translations messages.xlf (Serbian (Latin))
[ci skip]

* New translations django.po (Serbian (Latin))
[ci skip]

* New translations messages.xlf (Italian)
[ci skip]

* New translations messages.xlf (Italian)
[ci skip]

* New translations messages.xlf (Danish)
[ci skip]

* New translations messages.xlf (Italian)
[ci skip]

* New translations messages.xlf (Italian)
[ci skip]

* New translations messages.xlf (Polish)
[ci skip]

* New translations messages.xlf (Italian)
[ci skip]

* New translations messages.xlf (Luxembourgish)
[ci skip]

* New translations messages.xlf (Portuguese, Brazilian)
[ci skip]

* New translations messages.xlf (Chinese Simplified)
[ci skip]

* New translations messages.xlf (Turkish)
[ci skip]

* New translations messages.xlf (Swedish)
[ci skip]

* New translations messages.xlf (Slovenian)
[ci skip]

* New translations messages.xlf (Slovak)
[ci skip]

* New translations messages.xlf (Portuguese)
[ci skip]

* New translations messages.xlf (Dutch)
[ci skip]

* New translations messages.xlf (Hebrew)
[ci skip]

* New translations messages.xlf (Danish)
[ci skip]

* New translations messages.xlf (Finnish)
[ci skip]

* New translations messages.xlf (Greek)
[ci skip]

* New translations messages.xlf (German)
[ci skip]

* New translations messages.xlf (Czech)
[ci skip]

* New translations messages.xlf (Belarusian)
[ci skip]

* New translations messages.xlf (Spanish)
[ci skip]

* New translations messages.xlf (Romanian)
[ci skip]

* New translations messages.xlf (Russian)
[ci skip]

* New translations messages.xlf (French)
[ci skip]

* New translations messages.xlf (Serbian (Latin))
[ci skip]

* New translations messages.xlf (Arabic, Saudi Arabia)
[ci skip]

* New translations messages.xlf (German)
[ci skip]

* New translations messages.xlf (Polish)
[ci skip]

* New translations messages.xlf (German)
[ci skip]

* New translations messages.xlf (German)
[ci skip]

* New translations messages.xlf (German)
[ci skip]

* New translations messages.xlf (Spanish)
[ci skip]

* New translations messages.xlf (Spanish)
[ci skip]

* New translations messages.xlf (French)
[ci skip]

* New translations messages.xlf (Chinese Simplified)
[ci skip]

qcasey committed:
Merge pull request #902 from paperless-ngx/v1.7.1

v1.7.1

qcasey committed:
Append generated changelog to docs

qcasey committed:
Convert changelog to md

qcasey committed:
Add myst_parser for sphinx, changelog to md

qcasey committed:
Use checkout in action

qcasey committed:
Add links to usernames and PR #s

Co-authored-by: shamoon <4887959+shamoon@users.noreply.github.com>

tooomm committed:
fix typo

shamoon committed:
Merge pull request #941 from tooomm/patch-1

[Readme] Fix typo

shamoon committed:
Merge pull request #935 from qcasey/changelog-to-md

Convert Changelog to markdown, auto-commit future changelogs

tooomm committed:
small typos

tooomm committed:
one more

tooomm committed:
wording

shamoon committed:
Merge pull request #946 from tooomm/patch-1

[Docs] Fix 2 small typos

Berjou committed:
docs: fix some typos

shamoon committed:
Merge pull request #948 from Berjou/docs-typo

docs: fix some typos

dabenzel committed:
Update scanners.rst

dabenzel committed:
Update docs/scanners.rst

Co-authored-by: shamoon <4887959+shamoon@users.noreply.github.com>

shamoon committed:
Update shortcut link

qcasey committed:
Merge pull request #878 from dabenzel/patch-1

[documentation] added iOS Shortcut Link for Uploading PDF to paperless

qcasey committed:
Add myst-parser

shamoon committed:
Merge pull request #982 from qcasey/changelog-to-md

Add `myst-parser` to fix readthedocs

shamoon committed:
Update custom.css

[ci skip]

Commit:
Added Maxify MB5350 Documentation

shamoon committed:
Update changelog.md

qcasey committed:
Fix formatting, typo

qcasey committed:
Link to configuration values for Canon MFP

qcasey committed:
Merge pull request #984 from eingemaischt/canon-maxify-mb5350

Added Maxify MB5350 Documentation

qcasey committed:
Merge pull request #985 from paperless-ngx/fix-changelog-links

Fix links in changelog

tooomm committed:
update heading levels for v1.7.0

tooomm committed:
Link issues, capitalization and minor fixes

qcasey committed:
Merge pull request #990 from tooomm/patch-2

Docs: Fix headings and add links to PRs in changelog

martin31821 commented:
Any news regarding this PR? 
I'd love to use this with S3 as storage backend, the PR seems to work good for this purpose (tested with Postgres)

shamoon committed:
Add lsio to issue form

pReya committed:
fix: update scanner capability

The Brother ADS-A1700W does indeed support SFTP. I've just bought it, and set it up like this.

tribut committed:
Merge pull request #1148 from pReya/patch-1

fix: update scanner capability

shamoon committed:
Use any-of-labels for stalebot

[ci skip]

shamoon committed:
Update stale.yml

[ci skip]

shamoon committed:
Merge pull request #1110 from paperless-ngx/update-issue-form

a17t committed:
Merge commit '48dfbbebc654464026b0137c635262073c417292'

Conflicts:
	src/documents/management/commands/document_exporter.py
	src/documents/views.py

a17t commented:
> Any news regarding this PR? I'd love to use this with S3 as storage backend, the PR seems to work good for this purpose (tested with Postgres)

Hi Martin, 
sorry for the delay. I've been quite busy lately. Anyway the changes are also part of my personal paperless installation and seem to work fine. I just merged the latest main branch to my repository and will test that by upgrading my installation. If that works, I will change the state from WIP to "lets review it", so those changes can become part of paperless-ngx.
